<a href="https://colab.research.google.com/github/Carlos-Alberto-Antonio-G-Fuentes/Data_science_numpy/blob/main/Curso_de_Estad%C3%ADstica_con_Python_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***
# <font color=#56B4E9 size=10>CURSO DE ESTADÍSTICA</font>
***

<font color=#56B4E9 size = 5>*Preparando el ambiente*</font>

# <font color=#56B4E9>1 CONOCIENDO LOS DATOS</font>
***

## <font color=#56B4E9>1.1 Dataset del proyecto</font>
***

### Muestra de domicilios Colombia - 2018

Las investigaciones por muestras de domicilios realizadas anualmente, buscan encontrar característcias generales de la población, de educación, trabajo, rendimiento y otras, de acuerdo con las necesidades de información del país, tales como las características de migración, fertilidad, casamientos, salud, nutrición, entre otros temas. Estas muestras al pasar de los años consistuyen una herramienta importante para la formulación, validación y evaluación de políticas dirigidas al desarrollo socioeconómico y la mejora de las condiciones de vida en Colombia.


### Datos

Los datos fueron creados  de manera didáctica para este curso.

### Variables utilizadas

> ### Ingreso
> ***

Ingresos mensuales (en miles de pesos) del trabajo principal para personas de 10 años o más.

> ### Edad
> ***

Edad del entrevistado en la fecha de referencia en años.

> ### Altura
> ***

Altura del entrevistado em metros.

> ### Ciudad
> ***

Código de referência a 27 ciudades analizadas.


> ### Sexo
> ***

|Código|Descripción|
|---|---|
|0|Masculino|
|1|Femenino|

> ### Años de Estudio
> ***

|Código|Descripción|
|---|---|
|1|Sin estudios y menos de 1 año|
|2|1 año|
|3|2 años|
|4|3 años|
|5|4 años|
|6|5 años|
|7|6 años|
|8|7 años|
|9|8 años|
|10|9 años|
|11|10 años|
|12|11 años|
|13|12 años|
|14|13 años|
|15|14 años|
|16|15 años o más|
|17|No se sabe|
||No aplica|

> ### Color
> ***

|Código|Descripción|
|---|---|
|0|Indio|
|2|Blanco|
|4|Negro|
|6|Amarillo|
|8|Moreno|
|9|Sin declarar|

> ### Tratamiento a los datos
***

> Algunos de los tratamientos de datos más frecuentes son:
> 1. Eliminar las observaciones (líneas) con entradas de datos inválidos;
> 2. Eliminar observaciones donde hay datos perdidos (missing data);
> 3. Filtros própios de la investigación, por ejemplo: considerar solo las encuestas realizadas a la cabeza de familia (responsable por el domicilio).

### Importando pandas y leyendo el dataset del projecto

https://pandas.pydata.org/

In [2]:
import pandas as pd

In [9]:
datos = pd.read_csv("/content/datos.csv",encoding='latin1')

In [10]:
datos.sample(10)

Ciudad  Sexo  Edad  Color  Años de Estudio  Ingreso    Altura
49135      35     1    39      8                9     1200  1.791735
39591      32     0    25      2               12     1000  1.669457
17964      24     0    26      8                8       80  1.631557
75399      53     0    55      2               16     6000  1.722672
26316      29     0    26      2               12     1500  1.617572
68120      50     0    44      4               14      788  1.555899
64798      43     1    40      2               16     1200  1.679765
66589      43     1    24      2               13     2200  1.607963
51170      35     0    42      2               17     3000  1.647811
7636       15     0    44      8                6     1600  1.664628

Transcripción
Ustedes ya deben haber descargado, por ejemplo, nuestro archivo. Entonces, van a tener todos los documentos necesarios. Una vez que tengan el documento necesario y hayan abierto su notebook, estando en Google Collaboratory, esta es la visión que tendrán del notebook.

Comenzando por nuestro curso de estadística, vamos a conocer los datos, cómo los van a ver y cómo está organizado nuestro curso. Aquí, en la parte izquierda, en nuestra barra, verán que está todo el índice de nuestro curso. No hay forma de perderse. Tienen todas las aulas aquí y todos los tópicos que vamos a ir solucionando. Explicaremos cada uno de estos tópicos a medida que avancemos en el curso.

La primera parte que deben saber de nuestro dataset del proyecto es que tenemos una muestra de domicilio que fue hecha de manera anónima, inspirada en el Censo de Colombia de 2018. Tenemos muestras de domicilio con datos anonimizados, realizados internamente. ¿Para qué tenemos este dataset? Es importante que cada vez que tengamos una base de datos, como el dataset que estamos usando, entendamos qué debemos solucionar, qué preguntas queremos responder. Esta muestra de domicilio se realiza anualmente y hemos entrevistado a las personas que son cabeza de los hogares, es decir, los jefes o jefas de hogares.

En esta base de datos, hemos intentado dejarla de manera más fácil y comprensible para nuestros ejercicios. Nuestros datos incluyen variables como ingreso, edad, altura, ciudad, sexo, años de estudio, color, entre otras. Para cada una de las variables, hay una descripción. Por ejemplo, el ingreso mensual está en miles de pesos colombianos. Tenemos la edad de las personas entrevistadas, la altura, la ciudad, el sexo, los años de estudio, etc.

¿Cuál fue el tratamiento de datos? Es importante resaltar que al tratar datos, es necesario documentar lo que se ha hecho. Por ejemplo, hemos eliminado todas las observaciones o líneas con datos inválidos, datos nulos o missing data (datos vacíos). También hemos aplicado algunos filtros para facilitar el entendimiento y avanzar en nuestro curso.

Ahora, ¿qué debemos hacer? Primero, ¿cuál es la librería que usamos para manejar nuestro dataset del proyecto? Como ya han visto en algunos cursos, siempre usamos en Python la librería Pandas. Es una librería muy importante. ¿Cómo la importamos y comenzamos a trabajar? Siguiendo buenas prácticas, primero importamos la librería. Entonces, escribimos:

import pandas as pd
Copia el código
Le damos un sobrenombre, un nickname, que será pd. Es una buena práctica que la comunidad usa globalmente. Si desean usar otro sobrenombre, pueden hacerlo, pero sugerimos seguir las buenas prácticas recomendadas por la comunidad.

Importamos Pandas como pd y presionamos Shift + Enter. La primera vez puede demorar un poco porque se accede a los recursos de CPU en Google Collaboratory. Una vez hecho esto, verán un check indicando que Pandas ha sido importado como pd.

¿Cómo comenzamos a ver nuestra base de datos con todas estas variables? Primero, ¿cómo podemos ver los datos? Intentamos:

datos.head()
Copia el código
Si no hay datos, ¿cómo importamos nuestra librería? En la barra izquierda, hay un archivo con una flecha hacia arriba que dice Upload. Clicamos ahí. Ya tienen los datos descargados. Abran el archivo. Aparecerá un mensaje de advertencia importante. Cada vez que cierren su notebook o se ausenten, deberán ejecutar nuevamente su notebook y descargar la base de datos como se explicó.

Una vez descargado, asignamos a una variable llamada datos y llamamos a Pandas:

datos = pd.read_csv('ruta/del/archivo.csv')
Copia el código
Colocamos comillas simples o dobles para la ruta del archivo. Copiamos el camino y lo pegamos. Presionamos Shift + Enter.

¿Cómo leemos esta variable datos? Queremos imprimirla. Para ver las primeras cinco observaciones, usamos:

datos.head()
Copia el código
Pero lo que yo normalmente me acostumbré a hacer durante, varias observaciones o varios análisis que ya hice,es que a mí me gusta ver un sample.

datos.sample(10)
Copia el código
Esto nos permite ver 10 observaciones aleatorias y verificar que todo esté correcto.

Una vez hecho esto, podemos seguir con la segunda parte de nuestra aula. Cualquier duda que tengan, no duden en escribirnos. Los esperamos en la próxima aula. ¡Chao!

## <font color=#56B4E9>1.2 Tipos de datos</font>
***

### Clasificación de una variable
<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img001.png' width='70%'>

### Variables cualitativas ordinales

► Variables que pueden ser ordenadas o que responen algun tipo de jerarquía

In [13]:
datos.columns

Index(['Ciudad', 'Sexo', 'Edad', 'Color', 'Años de Estudio', 'Ingreso',
       'Altura'],
      dtype='object')

In [12]:
sorted(datos['Años de Estudio'].unique())

[np.int64(1),
 np.int64(2),
 np.int64(3),
 np.int64(4),
 np.int64(5),
 np.int64(6),
 np.int64(7),
 np.int64(8),
 np.int64(9),
 np.int64(10),
 np.int64(11),
 np.int64(12),
 np.int64(13),
 np.int64(14),
 np.int64(15),
 np.int64(16),
 np.int64(17)]

### Variables cualitativas nominales

► Variables que no pueden ser ordenadas ni responden alguna jerarquía.

In [14]:
sorted(datos['Sexo'].unique())

[np.int64(0), np.int64(1)]

In [15]:
sorted(datos['Color'].unique())

[np.int64(0), np.int64(2), np.int64(4), np.int64(6), np.int64(8)]

In [16]:
sorted(datos['Ciudad'].unique())

[np.int64(11),
 np.int64(12),
 np.int64(13),
 np.int64(14),
 np.int64(15),
 np.int64(16),
 np.int64(17),
 np.int64(21),
 np.int64(22),
 np.int64(23),
 np.int64(24),
 np.int64(25),
 np.int64(26),
 np.int64(27),
 np.int64(28),
 np.int64(29),
 np.int64(31),
 np.int64(32),
 np.int64(33),
 np.int64(35),
 np.int64(41),
 np.int64(42),
 np.int64(43),
 np.int64(50),
 np.int64(51),
 np.int64(52),
 np.int64(53)]

### Variables cuantitativas discretas

► Variables que representan un conteo donde los valores posibles forman un conjunto finito o numerable.


In [17]:
datos.Edad.max()

99

In [18]:
print(f"La edad mínima es {datos.Edad.min()} y la edad máxima es {datos.Edad.max()}")

La edad mínima es 13 y la edad máxima es 99


#### <font color='red'>Observación</font>
***
> La variable edad puede ser clasificada de tres formas diferentes:
> 1. <b>CUANTITATIVA DISCRETA</b> - cuando representa años completos (números enteros);
> 2. <b>CUANTITATIVA CONTINUA</b> - cuando representa la edad exacta, siendo representada por fracciones de años;
> 3. <b>CUALITATIVA ORDINAL</b> - cuando representa  intervalos de edad.

### Variables cuantitativas continuas

► Variables que representan un conteo o medición que asume valores en escalas continuas (números reales).

In [19]:
print(f"La Altura mínima es {datos['Altura'].min()} y la altura máxima es {datos['Altura'].max()}")

La Altura mínima es 1.339244614 y la altura máxima es 2.028496765


Transcripción
En la clase anterior, vimos cómo cargar nuestro dataset. Ahora, ya sabemos cómo hacerlo. No olvidemos, al finalizar la clase, volver a descargar nuestro archivo. Esto es importante para poder continuar. Podemos ir a "Runtime" y seleccionar "Run Before" para ejecutar los pasos anteriores y tener toda nuestra base de datos lista.

En esta clase, abordaremos conceptos básicos de estadística y tipos de datos. Es crucial entender la clasificación de las variables, ya que nos permite categorizar de manera efectiva. Esto es fundamental para crear gráficos como barras, histogramas o boxplots.

Las variables en estadística se dividen en dos grupos: cuantitativas y cualitativas. Esto ya lo hemos visto en clases anteriores. Las variables cualitativas se dividen en ordinales y nominales, mientras que las cuantitativas se dividen en discretas y continuas.

Las variables cualitativas ordinales tienen jerarquía y pueden ordenarse. Por ejemplo, en la Liga de la Justicia, Superman realizó 10 misiones, Batman 5 y Aquaman 3. Podemos decir que las misiones de Superman son superiores a las de Batman. En nuestro dataset, tenemos años de estudio, color, edad, sexo, ciudad, ingreso y altura. Algunas de estas variables, como sexo,No podemos decir que masculino, femenino, uno u otro es mayor que algo, no tienen jerarquía. Sin embargo, los años de estudio sí pueden ordenarse, podemos decir que la persona que ha sido entrevistada tiene 12 años de estudio, que es mayor de la persona que ha sido entrevistada de 5.

Para describir esto en Python, llamamos a nuestro dataset con datos['años de estudio'].unique(). Esto nos muestra los valores únicos de años de estudio. Podemos ordenarlos con sorted() para obtener un orden ascendente sorted(datos['años de estudio'].unique()). 1, 2, 3, 4, y por ahí va, y tenemos hasta 17 años de estudio. Entonces, de manera ordenada ya lo tenemos así.

Pasando para la parte de variables cualitativas nominales, son aquellas variables que no pueden ser ordenadas, no tienen ni una jerarquía, que es lo que vimos, por ejemplo, en nuestras variables cualitativas aquí, como ciudad, sexo y color, por ejemplo. Entonces, ¿qué hacemos aquí en nuestra base, por ejemplo? Aquí lo que tenemos que hacer, en este caso, ya tenemos nuestra línea de código. Vamos a copiar aquí, le damos Control C, y aquí, Control V, y en vez de ser años de estudio, vamos a colocar aquí sexo sorted(datos['Sexo'].unique()). Shift Enter, o 0, 1. La misma cosa vamos a hacer aquí para color sorted(datos['Color'].unique()), Shift Enter. Y, por último, vamos a hacer lo mismo aquí para ciudad sorted(datos['Ciudad'].unique()), Shift Enter.

Entonces, una cosa importante que ustedes tienen que saber de nuestro dataset es que, por ejemplo, ha sido hecho un tratamiento de datos para que nosotros, de alguna manera, codifiquemos la ciudad para, por ejemplo, no sé, Bogotá tiene el código 11, otra ciudad de Colombia tiene el código 53, por ahí, solo para, de una manera, ejemplificar mejor y nosotros podamos usar nuestra base de datos. Entonces, dicho esto, tenemos ya nuestras variables cualitativas y ya hemos conseguido usar y describirlas aquí en nuestro notebook. Sobre las variables cuantitativas discretas. Entonces, si ya sabemos que las variables cualitativas estamos hablando de una cualidad de, por ejemplo, una descripción de nuestro objeto, aquí estamos diciendo que nuestras variables cuantitativas son numéricas y podemos, básicamente, clasificarlas de manera numérica. En cuanto a las cualitativas no. Y las cuantitativas, disculpen, ellas se dividen, entonces, en discretas y continuas. Dentro de nuestro dataset, por ejemplo, tenemos alguna que tenemos, por ejemplo, la edad, tenemos el ingreso y tenemos la altura.

Y ahora, ¿cómo nosotros conseguimos clasificar aquí? Cuando estamos hablando de variables cuantitativas discretas, estamos diciendo que tiene un valor finito. ¿Qué significa, por ejemplo, tener un valor finito? Yo siento que la primera cosa que pensaron fue edad. Correcto. Edad, nosotros tenemos un nivel finito hasta donde nosotros vamos a vivir. Entonces, la edad es la misma cosa. No es algo continuo. ¿Entienden? Entonces, si edad, por ejemplo, es algo finito, ¿cómo hacemos para ver nuestra variable cuantitativa discreta aquí? Primero, vamos a llamar, como ya hicimos anteriormente, datos. Y aquí vamos a colocar Edad.unique() y hacemos así. Y aquí tenemos todos los tipos de edad que nosotros tenemos. Sin embargo, no es eso lo que yo quiero. Yo quiero ver, por ejemplo, la edad mínima o la edad máxima. En el caso, la edad mínima, datos.Edad.min() ha sido 13 en nuestro dataset. Y si cambiamos aquí, por ejemplo, en vez de ser min, colocamos un max, la edad máxima es 99.

Seguro ustedes se dieron cuenta que yo no vine a mi columna, por ejemplo, usando un corchete. En ese caso, por ejemplo, yo simplemente accedí aquí a mi columna colocando un punto. ¿Pero por qué yo hice eso? Porque, por ejemplo, aquí voy a mostrar un ejemplo mejor aquí de lo que yo estoy hablando. Cuando nosotros tenemos datos y, por ejemplo, queremos acceder a una columna, aquí, en este caso, va a ser edad, por ejemplo. Van a ver que va a ser el mismo resultado. Vamos a colocar MAX. Está bien. Sin embargo, nosotros teníamos una otra variable que él estaba datos. con espacio. Entonces, si él está con espacio, ¿qué es lo que nosotros recomendamos, que la comunidad recomienda y todo? Va a dar un error si nosotros colocamos esta variable aquí.

Vamos a dar CTRL-Z y vamos a copiar aquí. Si colocamos así, es una sintax inválida. ¿Por qué es inválida? Porque tiene espacio y también tiene, por ejemplo, símbolos, ñ, si tiene acento y todo. Él no va a entender. Entonces, ¿qué es lo que nosotros hacemos en este caso? Cuando, por ejemplo, está escrito el nombre de la columna de esta manera, es mucho más fácil nosotros, por ejemplo, entrar aquí. ¿Entienden? Nosotros acceder a las informaciones que está en nuestra base de datos. ¿Entienden? En nuestra columna. Sin embargo, por ejemplo, si nosotros colocamos años de estudio y le damos de esta manera datos['Años_de_Estudio'], tratamos el nombre de la columna así, de esta manera, por ejemplo, Python o Pandas va a entender, por ejemplo, que estamos accediendo a nuestra columna que tiene todo de una simple y de una buena vez. Entonces, él va a entender y nosotros podemos usar este método aquí de puntos. Cuando no, entonces, puntos nosotros usamos cuando tenga una palabra, cuando esté escrito de esta manera para acceder y sea mucho más fácil. Si no, puedes usar también corchetes y está todo bien. ¿Está bien?

Entonces, bien, volvemos aquí en la parte de edad. Entonces, vamos a sacar esto. Entonces, sabemos cuál es la edad máxima y cuál es la edad mínima. Pero, no sé si, por ejemplo, cuando nosotros vayamos a reportar esto aquí a nuestra jefa, a nuestro jefe, va a hacer esto de la manera correcta. ¿Qué es lo que usamos aquí para imprimir y dejarlo de una manera un poco más agradable de manera para reportar algo? Nosotros usamos una función que se llama print. Dentro de esta función print, nosotros tenemos el fprint. El fprint nos permite, por ejemplo, colocar entre llaves un cálculo que nosotros hemos hecho que va a ser esto de aquí. ¿Entienden? Entonces, aquí podemos decir, colocamos f, que va a ser una manera fstring. Nosotros estamos diciendo, usa esta función para que también podamos colocar un cálculo dentro de nuestro print.Que va a ser la edad mínima es, y abrimos la llave, y va a ser datos.Edad.min. La edad mínima es, y la edad máxima es, llaves de nuevo, datos.Edad.min. La edad mínima es 13, y la edad máxima es 99.

Entonces, aquí ya está un poco mejor explicado de lo que nosotros queremos hacer. Y ahí, bajamos aquí, tenemos las variables cuantitativas continuas. De nuevo, ustedes ya deben haber pensado, ¿qué es lo que nosotros cuando somos chiquitos y vamos creciendo a medida que, por ejemplo, vayamos aumentando nuestra edad? Es la altura. Entonces, continuamente estamos creciendo, disminuyendo, si estamos volviéndonos un poco viejos, de tamaño. Entonces, esto es continuo. ¿Entienden? Es algo continuo, un crecimiento continuo. Entonces, si nosotros tenemos, por ejemplo, nuestra variable altura, es importante que nosotros coloquemos esto.

Entonces, simplemente vamos a colocar aquí. Vamos a dar a Control-C. Vamos a dar a Control-V. Está. Entonces, colocamos que la altura mínima es, y aquí vamos, en vez de hacer edad. Vamos a colocar cochetes, y vamos a colocar Altura. Vamos a hacer diferente de lo que hicimos. ¿Entienden? Aquí también la altura máxima es. Y en vez de ser punto, vamos a colocar cochetes. Y va a ser aquí, Altura. Shift-Enter. Y listo. La altura mínima es de 1.33, y la altura máxima es de 2 puntos. Caramba, es una persona entrevistada bastante alta. Algo que ustedes han observado aquí, está como observación, está en rojo. La edad puede ser una variable bien tramposa cuando nosotros estemos usando nuestro dataset. ¿Por qué? Básicamente, la variable edad, como nosotros tenemos en nuestro dataset, está representada con números enteros. Entonces, por eso es que ella es una variable cuantitativa discreta. Sin embargo, cuando nosotros vayamos, por ejemplo, a un hospital, vamos a tener que nuestra variable edad está escrita con año, meses y día. Entonces, es la edad exacta. Entonces, si esto está así, normalmente la variable edad, en este caso, va a ser una variable cuantitativa continua. Por otro lado, también puede pasar que en nuestra edad, en algunas encuestas, en alguna base de datos, él esté representado por range. Está bien.

Entonces, como intervalos de edad. Entonces, puede ser que Erica tenga una edad entre 20 a 30 años. Gabriela tenga una edad entre 31 a 40 años. Entonces, si está de esta manera, está representada nuestra edad, entonces pasa a ser una variable cualitativa ordinal. ¿Está bien? Entonces, hemos visto todos nuestros tipos de variables. Hemos hablado también sobre por qué es importante categorizar a nuestras variables también de esta manera. ¿Está bien? Y dicho esto, entonces háganos ejercicio y en la próxima aula también nosotros vamos a ver un poquito más sobre la distribución de frecuencia y les voy a enseñar cómo hacer nuestras tablas de frecuencia. ¿Está bien? Un beso y los espero en la próxima clase. Chao, chao.

# <font color=#56B4E9>2 DISTRIBUCIÓN DE FRECUENCIAS</font>
***

El primer paso en un análisis es conocer el comportamiento de las variables involucradas en el estudio. Utilizando técnicas estadísticas como el análisis de las <b> DISTRIBUCIONES DE FRECUENCIAS </b> e <b> HISTOGRAMAS </b> podemos evaluar mejor cómo se distribuyen los fenómenos del estudio.



## <font color=#56B4E9>2.1 Distribución de frecuencias para variables cualitativas</font>
***

### Método 1

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.Series.value_counts.html

In [22]:
datos.Sexo.value_counts()

Sexo
0    53250
1    23590
Name: count, dtype: int64

In [24]:
datos.Sexo.value_counts(normalize=True)*100

Sexo
0    69.299844
1    30.700156
Name: proportion, dtype: float64

In [27]:
datos.Sexo.value_counts(normalize=True).round(4)*100

Sexo
0    69.3
1    30.7
Name: proportion, dtype: float64

In [30]:
frecuencia_sexo=datos.Sexo.value_counts()
frecuencia_sexo

Sexo
0    53250
1    23590
Name: count, dtype: int64

In [31]:
porcentaje_sexo = datos.Sexo.value_counts(normalize=True).round(4)*100
porcentaje_sexo

Sexo
0    69.3
1    30.7
Name: proportion, dtype: float64

In [42]:
distribucion_frec_cualitativa =pd.DataFrame({'Frecuencia':frecuencia_sexo,'Porcentaje %':porcentaje_sexo})

distribucion_frec_cualitativa

Frecuencia  Porcentaje %
Sexo                          
0          53250          69.3
1          23590          30.7

In [43]:
distribucion_frec_cualitativa.rename(index={0:'Masculino',1:'Femenino'},inplace=True)


In [44]:
distribucion_frec_cualitativa.rename_axis('Sexo',axis='columns',inplace=True)
distribucion_frec_cualitativa

Sexo       Frecuencia  Porcentaje %
Sexo                               
Masculino       53250          69.3
Femenino        23590          30.7

### Método 2

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.crosstab.html

In [49]:
sexo = {0: 'Masculino',
        1: 'Femenino'}

color = {0: 'Indigena',
         2: 'Blanco',
         4: 'Negro',
         6: 'Amarillo',
         8: 'Pardo',
         9: 'Sin declarar'}

In [50]:
frecuencia =pd.crosstab(datos['Sexo'],
            datos['Color'])
frecuencia

Color    0      2     4    6      8
Sexo                               
0      256  22194  5502  235  25063
1      101   9621  2889  117  10862

In [51]:
frecuencia.rename(index=sexo,inplace=True)
frecuencia.rename(columns=color,inplace=True)
frecuencia

Color      Indigena  Blanco  Negro  Amarillo  Pardo
Sexo                                               
Masculino       256   22194   5502       235  25063
Femenino        101    9621   2889       117  10862

Transcripción
Lo que vimos en la clase pasada incluyó cómo cargar nuestro documento, archivo o base de datos cuando estemos utilizando nuestro notebook. También discutimos conceptos fundamentales que son cruciales a la hora de clasificar nuestros tipos de datos. Además, vimos cómo utilizar Python para clasificar nuestras variables cualitativas y cuantitativas.

Para reforzar lo aprendido, es útil realizar un "run before", lo que implica volver a cargar la base de datos y repetir todo lo que hemos hecho hasta el momento. Esta práctica asegura que siempre estemos acostumbrados a realizar las operaciones necesarias de forma sistemática.

En la clase de hoy, Vamos a comenzar a ver nuestras distribuciones de frecuencia. ¿Qué es lo que esto significa?Vamos a ver cómo crear una tabla de frecuencia primeramente para nuestras variables cualitativas.¿Por qué usamos distribuciones de frecuencia? Porque es súper importante para entender el comportamiento de nuestra base de datos.

Por ejemplo, queremos saber cómo se comportan las personas entrevistadas del sexo femenino y del sexo masculino. ¿Cuál es el tipo de ingreso? ¿Tienen más de cuántos salarios mínimos? Para eso sirve nuestra distribución de frecuencia. ¿Está bien?

Comenzando con nuestras variables cualitativas, cada vez que vayamos a usar algunos métodos o funciones fundamentales para la creación de algunas tablas, siempre habrá una documentación. Sugiero que lean la documentación cuando vayan a desarrollar un proyecto personal. Puede suceder que la versión que estamos usando ahora sea diferente a la versión que vean más adelante cuando estén viendo este curso. Siempre estará disponible la documentación. Ahí estará el redireccionamiento.

Vamos a comenzar a usar la función ValueCounts. Es súper importante, ¿está bien? Vamos a usar la distribución de frecuencia para nuestras variables cualitativas. Lo primero que tenemos que hacer es identificar nuestras variables cualitativas. Vamos a comenzar con la variable sexo. Nuestra base de datos tenía la variable sexo codificada como 0 y 1, donde 0 era masculino y 1 era femenino. ¿Cómo accedemos a estos datos?

Primero, accedemos a datos. Vamos a usar "punto", porque no tiene espacio,Sexo, y luego colocamos la función "ValueCounts". Esto nos mostrará las frecuencias de personas femeninas y masculinas entre nuestros entrevistados. Entonces, nuestra base de datos muestra que 53.250 son hombres y 23.590 son mujeres. Tenemos esto. ¡Belleza! Ya tenemos un paso adelante. Ahora queremos ver lo mismo, pero en porcentaje. ¿Cómo hacemos para ver en porcentaje? Vamos a copiar esto, "Control-C". Dentro de nuestra función ValueCounts, tenemos un parámetro llamado Normalize. Está dentro de la documentación de pandas. Le daremos el valor True, porque por defecto la función trabaja con valores absolutos. Una vez que le damos "True", le estamos diciendo que nos lo muestre en forma de porcentaje. Así queda mejor también. ¿Ves que queda así? Por ejemplo, si damos Shift-Enter va a estar 0.69, 0.30. ¡Belleza! Pero vamos a colocar el por * 100, para que quede mejor. Y aquí tenemos que tenemos 69% y por ahí va.

Solo que yo puedo usar, por ejemplo, otra función que se llama round. Y esta función me va a redondear mis valores. Entonces, vamos a ver cómo funciona. round. Por defecto, él ya redondea todo para acumulativo. Entonces, aquí le vamos a dar unos 4 decimales. Y aquí se coloca 69.3 masculino y 30.7 femenino. ¿Está bien? Entonces, vamos a continuar. Aquí ya tenemos la parte de valores absolutos y los porcentajes.

Entonces, vamos a comenzar a ver nuestra frecuencia. Vamos a comenzar a usar nuestra tabla de frecuencia. Entonces, si ya tenemos datos en ValueCount y la parte de porcentajes, vamos a depositar eso en una variable. Vamos a depositar esta primera línea que hemos hecho aquí. ¿Está bien? Y depositar en esta variable que es frecuencia. ¿Entienden? Vamos a dar Shift Enter. Y cada vez que depositamos eso, atribuimos en una variable. Lo que estamos haciendo es colocar en esa caja, pero no estamos diciendo, muéstrame esa caja. ¿Qué es lo que tiene dentro de esa caja? Entonces, para eso, siempre que queremos ver qué hay, colocamos una línea después y solo llamamos a nuestra variable.

Ahora tenemos nuestra frecuencia y vamos a crear una nueva variable que será porcentaje. Vamos a dejar esto aquí padronizado para que sea frecuencia_sexo. ¿Está bien? Para que quede más específico. Está bien, cada vez que queramos buscar eso. Entonces, será Shift Enter. Aquí será porcentaje_sexo. Y tendremos esto de aquí que hemos hecho. Entonces, Control C. ¿Está bien? Y vamos a mandar a imprimir de nuevo para saber qué está dentro. Shift Enter. Y tenemos 69.3 y 30.7. ¿Está bien? Entonces, ya tenemos nuestras variables hechas. Y ahora queremos ver nuestra distribución de frecuencia. Nuestra tabla. ¿Cómo hacemos para ver nuestra tabla?

Para crear nuestra tabla, necesitamos crear un nuevo DataFrame. Y para crear un nuevo DataFrame, necesitamos llamar a nuestro amigo Pandas. Para eso existe una función que decimos así: Pandas, ¿cómo llamamos a Pandas? pd. Crea para mí, por favor, un DataFrame. ¿Está bien? Y aquí vamos a abrir nuestros paréntesis. Y le vamos a colocar, de alguna manera, por ejemplo, para que aparezca. Cuando estemos creando nuestro DataFrame, vamos a crear las columnas que queremos traer. Estamos diciendo que trae, por favor, una columna que se llame Frecuencia. Le llamaremos Frecuencia aquí. Frecuencia. Y esa columna será nuestra variable frecuencia_sexo. ¿Entienden? Y además, no solo quiero esa columna, también quiero otra que se llame Porcentaje. ¿Correcto? Porcentaje. Y esa columna, Porcentaje, será la variable que hemos creado, que será porcentaje_sexo. ¿Entienden? Y ahí le damos Shift-Enter, o como él creó nuestra tabla, nuestro DataFrame, de nuevo. ¿Quedó bien? Pero, y si llamamos de nuevo nuestra tabla, no va a ver cómo llamarla. Entonces, aquí vamos a atribuir esto aquí, en una nueva variable, que va a ser dist_frec_cualitativa. Shift-Enter, solamente para que actualice. Y vamos a llamar aquí nuestra variable. Aquí, dist_frec_cualitativa. Mismo resultado que vimos anteriormente. ¿Está bien?

Y ahí, ¿qué es lo que yo quiero también? Oh, no sé si se han dado cuenta que... ¡Belleza! Sexo es 0 y 1. Pero si ustedes van a, por ejemplo, mostrar esto para una persona que es líder de ustedes, él no va a entender que es 0 y 1. Entonces, vamos a tener que renombrar nuestros valores aquí. ¿Y cómo nosotros renombramos? Primero, nosotros tenemos que pensar que esta parte de aquí es un Axis, ¿no? Entonces, aquí está dentro del Axis, que va a ser de columna. Entonces, lo que queremos nosotros cambiar va a ser esta parte de aquí. Entonces, ¿cómo hacemos para cambiar esto? Primero, traemos a nuestra variable, que va a ser dist_frec_cualitativa. Y ahí nosotros vamos a dar, ¿qué? Método que va a ser rename_axis. Renombrea, por favor, mi Axis aquí. ¿Dónde? Sexo, ¿no? Que está escrito sexo. Sea lo siguiente. Va a estar en el axis, columna. Puede ser Index y en este caso columna, pero en nuestro caso vamos a colocar columna. ¿Entienden? Y va a estar acá, columns. ¿Está bien? Y ahí para que realmente pase y, por ejemplo, cambie de verdad de verdadcita nuestro DataFrame, nosotros tenemos que usar este parámetro que va a ser inplace = True. ¿Por qué? Si nosotros simplemente dejamos así, vean el cambio. Si yo traigo de nuevo nuestra variable que estábamos hablando anteriormente, ni cambió, ¿no? Está aquí. Está 0 y 1, 0 y 1, por ejemplo. Por ejemplo, si nosotros queremos cambiar aquí nuestra variable sexo para lo que nosotros hemos hecho, entonces lo que nosotros tenemos que hacer aquí es colocar el parámetro inplace = True. Entonces le colocamos aquí inplace = True. Ahora sí. Distribución de frecuencia cualitativa. Y ahora sí damos Shift-Enter.

Pareció un error aquí. Déjame ver. ¡Ah! Pero ¿qué era lo que yo estaba queriendo cambiar? Profe, usted no vio, pero ve ya, se está colocando en sexo y todo, pero él no está cambiando el 0 y 1. ¿Por qué no está cambiando? Porque básicamente nosotros no le hemos dado un rename al índex. Perfecto. Quien se dio cuenta de eso está a 100 puntos. ¿Por qué? dist_frec_cualitativa. Y aquí le vamos a dar rename. ¿Y qué era lo que yo les decía a ustedes? Entonces aquí, esta parte de aquí que es nuestro axis de columnas y todo, esto de aquí no es una columna, si ustedes se dan cuenta. Hasta por ejemplo, cuando ustedes ven aquí cómo es que está alineado, esto aquí es un índice. Y estos ya son las columnas. Entonces si nosotros queremos renombrar esto de aquí, tenemos que usar rename. Aquí le vamos a dar rename y aquí abrimos paréntesis y va a ser index. Y aquí cuando nosotros abramos el índex, vamos a abrir un diccionario donde 0 va a ser nada más que Masculino. ¿Entienden?Y aquí va a ser, por ejemplo, 1 nada más es que Femenino.Masculino y femenino.¿Está bien?Y de nuevo, si queremos modificar eso de manera de verdad, de verdadcita, tenemos que dar el parámetro inplace = True.¿Entienden?Y ahora sí, shift, enter.Si nosotros damos de nuevo la línea en nosotros que ya hemos trabajado hablando de distribución de frecuencia, ya cambiamos el índex y estamos diciendo a dónde queremos atribuir esto. Queremos atribuir en el eje que es sexo y el eje es la columna que estamos viendo. Entonces va a ser un inplace = True.Y aquí, por ejemplo, le damos shift enter.Ahora sí, tenemos masculino y femenino.¿Entienden?

Entonces, una vez que tenemos la parte de masculino y femenino, ya está correctamente nuestra tabla.Sin embargo, tenemos otro método que nos permite ver mejor.¿Cómo lo conseguimos?Aquí de nuevo tenemos una documentación que podemos usar, que es utilizando el crosstab.¿Y qué significa el crosstab?Nosotros podemos, de alguna manera, ver dos variables al mismo tiempo.Por ejemplo, estamos viendo color y sexo, o podríamos ver otra variable que también sea importante.Aquí ya hice un trabajo para ustedes, que es similar a lo que hemos hecho, solo que creé un diccionario que clasifica masculino y femenino para sexo, y para color con categorías como indígena, blanco, negro, amarillo, etc.¿Entienden?Entonces, ¿cómo empezamos a usar el método del crosstab?

Primero, colocamos la frecuencia.No, no depositamos una variable.Voy a mostrar cómo lo hacemos.Llamamos a nuestro amigo pandas, llamamos a la función que queremos ver, que es el crosstab, y aquí entramos. Primero, los datos base de datos, columna .Sexo, base de datos datos, columna Color, y traemos la columna de color. ¿Entienden?Y aquí, shift enter.¡Oh, qué maravilla!Tenemos el sexo masculino, femenino, y las categorías como indígena, blanco, etc.Entonces, le vamos a atribuir a la variable frecuencia.Aquí va a ser frecuencia, ¿entienden?Y de nuevo, shift enter para actualizar.¿Está bien?Entonces ya tenemos nuestra variable frecuencia con el método del crosstab, ¿entienden?Ahora vamos a renombrar todo esto.¿Por qué?Si imprimimos esto de aquí con shift enter, no se entiende.Por ejemplo, tu líder no va a entender qué es 0, 2, 4, 6, 8.Entonces vamos a usar el diccionario que ya creamos anteriormente.¿Cómo renombramos?

De nuevo, frecuencia, frecuencia.rename.Renombra el índice que es sexo, ¿está bien?Y aquí usamos nuestro diccionario.Recuerden, este diccionario de aquí.Entonces, índice, y para que realmente cambie, le damos inplace = True.Luego, llamamos a nuestra variable frecuencia y renombramos las columnas.Entonces, columns, que va a ser nuestro diccionario de color.Y de nuevo, para que realmente pase el cambio, le damos inplace = True.Por último, mandamos a ejecutar.Vamos a ver si pasó.

Opa, is not defined.¿Por qué no está definido?No está definido porque no he ejecutado el diccionario.Entonces, shift enter para ejecutar nuestro diccionario.¿Ven que tiene un check?Entonces ya está. Normalmente, este error de "name sexo is not defined" es porque no hemos ejecutado el diccionario.No existe esa variable.Entonces, está diciendo: no has definido esto. Una vez que ejecutamos el diccionario, va a existir. Y aquí, shift enter, y colocamos nuevamente. ¡Oh, check, salió bien! Y ahora veremos la frecuencia. ¡Oh, maravilla!Entonces, tenemos que dentro de nuestra base de datos, 256 son masculinos del color indígena. Tenemos que 22.194 son blancos, negros, y por ahí va.

El próximo ejercicio que a mí me gustaría ver es el siguiente: digamos que yo soy tu líder y te digo, "Ah, genial, tenemos masculino y femenino. Pero ahora me gustaría ver el porcentaje. ¿Cuál es la proporción de masculino para indígena? ¿Masculino para blanco? ¿Masculino para las personas que son negras, amarillas y pardas?" Y así sucesivamente.Te dejo ese ejercicio. Y una vez lo hayas terminado, en la próxima clase te voy a explicar cómo ha sido desarrollado este ejercicio. ¿Está bien?Te veo en la próxima clase, donde también vamos a ver la distribución de frecuencia para las variables cuantitativas. ¿Está bien?Chao, chao.

##Ejercicio 1

In [54]:
porcentaje = pd.crosstab(datos.Sexo,
                         datos.Color,
                         normalize=True).round(4)*100
porcentaje.rename(index = sexo,inplace = True)
porcentaje.rename(columns = color,inplace = True)
porcentaje

Color      Indigena  Blanco  Negro  Amarillo  Pardo
Sexo                                               
Masculino      0.33   28.88   7.16      0.31  32.62
Femenino       0.13   12.52   3.76      0.15  14.14

##Ejercicio
En el mundo del análisis de datos, a menudo es necesario explorar la relación entre dos variables categóricas. El código que hemos visto utiliza la función pd.crosstab de Pandas para crear una tabla de contingencia, que nos permite examinar cómo se distribuyen los valores de una variable (Ingreso en este caso) en función de las combinaciones de otras dos variables categóricas (Sexo y Color).

Veamos en detalle cómo funciona este código:

In [52]:
porcentaje = pd.crosstab(datos.Sexo,
                         datos.Color,
                         aggfunc = 'mean',
                         values = datos.Ingreso)
porcentaje.rename(index = sexo,inplace = True)
porcentaje.rename(columns = color,inplace = True)
porcentaje

Color         Indigena       Blanco        Negro     Amarillo        Pardo
Sexo                                                                      
Masculino  1081.710938  2925.744435  1603.861687  4758.251064  1659.577425
Femenino   2464.386139  2109.866750  1134.596400  3027.341880  1176.758516

Explicación paso a paso:

pd.crosstab(datos.Sexo, datos.Color...): Esta es la función principal que crea la tabla de contingencia.
Los dos primeros argumentos (datos.Sexo y datos.Color) especifican las dos variables categóricas que se cruzarán.
aggfunc = 'mean': Indica que queremos calcular la media de datos.Ingreso para cada combinación de Sexo y Color. Podríamos usar otras funciones de agregación como sum, count, etc., dependiendo del análisis que queramos realizar.
values = datos.Ingreso: Especifica la variable numérica cuyos valores se agregarán en cada celda de la tabla.
porcentaje.rename(index = sexo, inplace = True): Esta línea renombra las etiquetas del índice (filas) de la tabla de contingencia utilizando los valores del diccionario sexo. El argumento inplace = True modifica la tabla directamente en lugar de crear una nueva.
porcentaje.rename(columns = color, inplace = True): De manera similar, esta línea renombra las etiquetas de las columnas de la tabla utilizando los valores del diccionario color.
porcentaje: Finalmente, se muestra la tabla resultante.
¿Qué nos muestra esta tabla?

La tabla resultante nos mostrará la media de Ingreso para cada combinación de Sexo y Color. Esto nos permite comparar, por ejemplo, si hay diferencias significativas en los ingresos promedio entre hombres y mujeres de diferentes etnias.

Te animo a experimentar con diferentes funciones de agregación y variables en pd.crosstab para descubrir patrones interesantes en tus datos. ¡El análisis bivariado es una herramienta poderosa para comprender mejor las relaciones entre variables en tus conjuntos de datos! ;))

## <font color=#56B4E9>2.2 Distribución de frecuencia para variables cuantitativas (clases personalizadas)</font>
***

### Paso 1 - Especifique los límites de cada clase

Utilice la siguiente clasificación:

<b>A</b> ► Más de 20 SM

<b>B</b> ► De 10 a 20 SM

<b>C</b> ► De 4 a 10 SM

<b>D</b> ► De 2 a 4 SM

<b>E</b> ► Hasta 2 SM

donde <b>SM</b> es el valor del salario mínimo en ese momento. En nuestro caso <b>788 mil pesos colombianos</b> (2018):

<b>A</b> ► Más de 15.760

<b>B</b> ► De 7.880 a 15.760

<b>C</b> ► De 3.152 a 7.880

<b>D</b> ► De 1.576 a 3.152

<b>E</b> ► Hasta 1.576


In [56]:
datos.Ingreso.min()

0

In [55]:
datos.Ingreso.max()

200000

In [57]:
clases = [0,1576,3152,7880,15760,200000]

In [58]:
label = ['E','D','C','B','A']

### Paso 2 - Crear la tabla de frecuencias

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.cut.html

In [60]:
frecuencia = pd.value_counts(
pd.cut(x = datos.Ingreso,
                    bins = clases,
                    labels = label,
                    include_lowest = True)
)
frecuencia

/tmp/ipython-input-1452934977.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  frecuencia = pd.value_counts(


Ingreso
E    49755
D    16700
C     7599
B     2178
A      608
Name: count, dtype: int64

In [64]:
frecuencia_porcentaje = pd.value_counts(
                    pd.cut(x = datos.Ingreso,
                    bins = clases,
                    labels = label,
                    include_lowest = True),
                    normalize =True).round(4)*100
frecuencia_porcentaje


/tmp/ipython-input-3080532994.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  frecuencia_porcentaje = pd.value_counts(


Ingreso
E    64.75
D    21.73
C     9.89
B     2.83
A     0.79
Name: proportion, dtype: float64

In [65]:
distribucion_frec_cuantitativa =pd.DataFrame({'Frecuencia':frecuencia,'Porcentaje %':frecuencia_porcentaje})

distribucion_frec_cuantitativa

Frecuencia  Porcentaje %
Ingreso                          
E             49755         64.75
D             16700         21.73
C              7599          9.89
B              2178          2.83
A               608          0.79

In [66]:
distribucion_frec_cuantitativa.sort_index(ascending=False,inplace=True)

distribucion_frec_cuantitativa

Frecuencia  Porcentaje %
Ingreso                          
A               608          0.79
B              2178          2.83
C              7599          9.89
D             16700         21.73
E             49755         64.75

## <font color=#56B4E9>2.3 Distribución de frecuencia para variables cuantitativas (clases de amplitud fija)</font>
***

### Importando bibliotecas

http://www.numpy.org/

In [67]:
import numpy as np

### Paso 1 - Difiniendo el número de clases

#### Regla de Sturges

# $$k = 1 + \frac {10}{3}\log_{10}n$$

In [68]:
datos.shape

(76840, 7)

In [69]:
datos.shape[0]

76840

In [71]:
k = 1 + (10/3) * np.log10(datos.shape[0])
k

np.float64(17.285291187298853)

### Paso 2 - Crear la tabla de frecuencias

In [ ]:
frecuencia_sturges = pd.value_counts(
pd.cut(x = datos.Ingreso,
                    bins = 17,
                    include_lowest = True)
)
frecuencia_sturges

Transcripción En la clase anterior, entonces, ya hemos visto cómo hacer nuestra tabla de frecuencia para las variables cualitativas. Entonces, ya hemos visto, por ejemplo, la distribución de frecuencia exactamente usando nuestra variable sexo y también hemos conseguido combinar en el método dos variables, que va a ser sexo y color, y un ejercicio que les había propuesto que ustedes hagan.

Además de los valores absolutos, por ejemplo, y hasta le hemos modificado y renombrado aquí los nombres de nuestras columnas, nuestras variables y todo, nosotros queríamos ver, en este caso, voy a fingir que soy la líder de ustedes, y les dije, ah, yo quiero ver también cómo se ve la proporción, además de los valores absolutos, cómo hacer.

Entonces, ustedes ya han visto en un ejercicio anterior que nosotros hemos visto cómo crear la parte de porcentaje, entonces, nada más es copiar la misma pd.crosstab que nosotros ya habíamos visto aquí en valores absolutos, y aquí nosotros, ¿se acuerdan que nosotros teníamos la parte, un parámetro de normalize = true?

Es adicionar este parámetro, normalize = true, y de nuevo, reemplazar el nombre del sexo con nuestro diccionario que hemos creado anteriormente, renombrar, por ejemplo, nuestras columnas con nuestro diccionario de color, y no se olviden nunca, por ejemplo, si queremos que de verdad se hagan estos cambios, nosotros tenemos que usar el parámetro inplace = true, y por último, le damos shift-enter aquí, y están aquí las proporciones que hemos visto.

¿Está bien? Entonces, es súper importante que ustedes hagan esto. ¿Qué es lo que vamos a ver en esta aula ahora?

Vamos a ver nuestras variables y crear nuestras tablas de frecuencias para variables cuantitativas, ya que hemos visto las cualitativas.

Entonces, el primero paso que nosotros vamos a hacer, dado que vamos a usar nuestra variable cuantitativa, que ahora va a ser nuestra variable Ingreso, nosotros, como primer paso, es especificar los límites de cada clase. Entonces, nosotros tenemos que el salario mínimo, por ejemplo, en el momento que nosotros estamos viendo la base de datos en 2018, como un dato bien así, que hemos usado ahí dentro de nuestra tabla del censo del 2018, que fue usado en Colombia, fue la última vez que se ha hecho un censo, el salario mínimo ahí era de 788.000 pesos colombianos.Dicho esto, si 788.000 pesos colombianos eran en 2018, lo que nosotros tenemos que ver, por ejemplo, es... Ah, para categorizar nuestras clases tenemos que A es más de 20 salarios mínimos, B es de 10 a 20 salarios mínimos, C de 4 a 10, 2 a 4 y así sucesivamente.

Entonces, si es 788.000, es 788.000 veces 20, 788.000 veces 4, 10 y por ahí va, hasta que nosotros tengamos aquí nuestros valores que nosotros queremos.Entonces, tenemos que la categoría A va a ser de más de 15.760, B de 7.880.000 a 15.760 y así sucesivamente con ese cálculo, ¿entienden?

Entonces, ¿qué es lo que vamos a hacer ahora? Primero, vamos a explorar un poco más nuestra variable de ingreso, entonces va a ser datos.Ingreso.min(), ¿cuál es el salario mínimo? Cero, está bien.

Y ahí nosotros queremos saber cuál es el salario máximo de la persona que ha sido entrevistada. Va a ser datos, de nuevo, entrar a nuestro dataset, datos.Ingreso.max(), el valor máximo. Y tenemos que el valor máximo aquí es 200.000 pesos colombianos. ¡Belleza! Tenemos eso.

Solo que ahora vamos a comenzar a usar nuestras clases y nuestros labels. Entonces, ¿cuál son nuestras clases? Nuestras clases son estos, ¿correcto?

Entonces, clases = va a ser igual a, abrimos aquí, corchetes, y va a ser de 0 a 1.576, va a ser de 1.576 a 3.152, ¡Opa! 3.152, después de 3.152 a 7.880, y así sucesivamente hasta llegar a nuestro valor máximo que es 200.000. Entonces tenemos eso de aquí, Shift-Enter, y ahora vamos a colocar nuestros labels. ¿Qué son nuestros labels? Nada más que los nombres de las clases que él va a acabar haciendo. Entonces, imaginen que cada una de estas clases es un intervalo, ¿me entienden? Entonces, para este intervalo, el label será el nombre A. Y para este intervalo, será el nombre B. Esto es lo que significa label. Colocaremos aquí nuestra variable label, que será la siguiente. Va a ser ['E', 'D', 'C', 'B', 'A'] , de acuerdo al orden que hemos establecido, ¿me entienden? ¿Cómo podemos ver más adelante esa tabla de frecuencia? Solo para que vean que está depositado de manera adecuada. Imprimiremos nuestra variable clase, y la serie también, ¿está bien?

¿Qué haremos en nuestro segundo paso? Crear nuestra tabla de frecuencia, ¿por qué no? Recuerden que siempre deben leer la documentación. Para usar una tabla de frecuencia para datos cuantitativos, vamos a usar una función muy importante: pd.cut. ¿Qué es pd.cut? Es parte de pandas punto cut. Si recuerdan, le hemos colocado un sobrenombre que es pd, entonces será pd.cut, ¿entienden? Aquí crearemos nuestra variable, que llamaremos frecuencia, y en esa frecuencia colocaremos… Pandas, por favor, traeme esta función que vamos a usar, que será cut. Antes de crear, por ejemplo, pd.cut, queremos ver los valores absolutos. Para ver los valores absolutos, lo que haremos es lo siguiente.

Primero, pd.cut, y estamos diciendo que x será nuestra base de datos. Esa base de datos será simplemente datos. ¿Cuál es la variable que estamos viendo? En este caso será Ingreso, que está escrito así, correcto. También queremos añadir un parámetro que serán nuestros bins, que son las clases, los intervalos que hemos creado. Los bins serán nuestras clases. Label, que aquí es muy redundante, pero… déjame ver, hemos creado aquí labels, ¿está bien? Un parámetro muy importante aquí es que queremos incluir el valor de cero, y para eso usamos include_lowest. Por defecto, viene como falso, pero aquí lo pondremos como True. Damos un shift + enter, y aquí veremos la frecuencia, y ya está. Está correcto, y vemos todos los valores que estamos revisando.

Entonces, ¿cuántos E tenemos? ¿Cuántos C tenemos? ¿Cuántos A tenemos? Para hacer eso, como les mencioné en la teoría, ahora necesitamos usar pandas.value_counts. ¿Cómo añadir este parámetro también? Se hará aquí, adelante de este pd.cut, arriba en realidad, será pd.value_counts. Abrimos paréntesis, y aquí, dentro de ese paréntesis, estará todo lo que hemos hecho con pd.cut. Cerramos el paréntesis. Una cosa importante es verificar cuántos paréntesis tenemos. Muchas veces, un paréntesis que no se ve correctamente causa un error, y eso es muy común. Así que es crucial que lo revisen. Damos shift + enter, y ahora sí, aquí se da frecuencia, oh, voilà. Entonces, tenemos que en E hay 49.755 personas en la categoría E. En la categoría D, son 16.700, y así sucesivamente, ¿entienden? Entonces, esto aquí lo vamos a dejar solo para organizar, en este código, para que esté mejor.

Ahora que tenemos nuestros valores absolutos, ¿qué es lo que queremos ver también? ¿Por qué no ver porcentajes? Vamos a copiar este mismo código, y ¿se acuerdan que dentro de value_counts nosotros teníamos un parámetro que era para normalizar? Es lo mismo, solo que tiene que estar dentro de los paréntesis de value_counts, no dentro de los de pd.cut. Entonces, después que hemos venido con pd.cut, aquí, después de ese paréntesis, damos coma, y aquí va a ser, abajo, normalize, normaliza para mí, va a ser True, y aquí va a ser * 100. En vez de ser frecuencia, vamos a cambiar para porcentaje nuestra variable, y aquí va a ser porcentaje. ¡Oh, qué bonito! Aquí tenemos, entonces, la categoría E, que representaba 49.755, que eso en proporciones sería el 65% de nuestra base. Aquí vamos a dar un round(4), para dejarlo un poco mejor en el entendimiento. Vamos a colocar, 64.75 de nuestra base en la categoría E, y así seguimos. Ya tenemos nuestra tabla con frecuencia y proporción.

Ahora, ¿cómo juntamos estas dos informaciones en una tabla en conjunto? Ya hemos visto esto anteriormente, donde hemos juntado estas dos informaciones. Voy a intentar traer este código, y aquí vamos a crear nuevamente, en vez que sea distribución de frecuencia cualitativa, va a ser cuantitativa, personalizada, porque hemos personalizado nuestros diccionarios. En vez de ser nuestra columna frecuencia, va a ser solo frecuencia, que fue lo que hemos visto. Y en porcentaje, en vez de porcentaje, va a ser porcentaje. Damos Shift+Enter, y aquí está nuestra dist_frec_personalizada. Solo que queremos ver, por ejemplo, de una manera organizada. ¿Cómo conseguimos ver eso? Existe una función que sería sort_index, y aquí sería ascending, o sea, de manera ascendente, y va a ser false. Shift+Enter, y voilà. Ahora, tenemos que él está colocando por Ingreso, organizando por el índice. Entonces, A, B, C, D, E, y aquí tenemos la frecuencia y el porcentaje de cada una de ellas.

Ahora, tenemos esta distribución de frecuencias personalizadas cuantitativas. Si queremos algo que tenga una amplitud fija, existe una teoría que se llama la regla de Sturges. Para eso, ¿qué es lo que sugiere? Imaginen que están organizando su biblioteca en casa, solo que tienen una inmensa cantidad de libros, y necesitan saber cuántos estantes van a querer para su biblioteca. Para saber este número de estantes, hacen la regla de Sturges. Él va a definir correctamente un estante donde tus libros no queden ni muy apretados ni estén danzando ahí en su estante. Básicamente, lo que hace es este cálculo, ¿entienden?Y ahí él va a definir exactamente cuál es la regla de clases que ustedes van a necesitar usar.Entonces, por eso es que se llama de amplitud fija. Para nosotros hacer este cálculo de manera manual, vamos a usar ahora una nueva librería que va a ser un biblioteca, que va a ser NumPy.

Entonces, vamos a descargar esto de aquí, vamos a importar. Entonces, importa para mí, por favor, import Numpy as np. De nuevo, le estamos dando un sobrenombre que va a ser np, y por la comunidad y todo, siempre le damos np. Entonces, sigan eso por padrón, pero si ustedes quieren colocar, no sé, MyNP o alguna cosa, pueden.Pero para proyectos y portafolios, sugiero que siempre sea algo de la comunidad, ¿entienden?Hasta, por ejemplo, de manera universal, si quieren mostrar su portafolio.Ahí nosotros tenemos eso.Ahora vamos a pasar a calcular.¿Y cómo calculamos nuestra, aquí, nuestra regla de Storages para tener nuestra amplitud fija? Primero, nosotros tenemos que ver el tamaño de nuestra base de datos.¿Cómo calculamos el tamaño? datos, traemos nuestra base de datos, y ahí vamos a querer saber el tamaño.Entonces shape, Shift-Enter .Entonces, tenemos aquí que nuestra base de datos contiene 76.840 registros o líneas, y tiene 7 columnas.Pero yo no quiero las columnas, porque esto aquí no me dice nada cuando yo vaya a hacer el cálculo.Ahí, para eso, voy a colocar aquí, corchetes, y voy a localizar, digamos así, para que venga apenas el primer valor datos.shape[0].Shift-Enter, O.Y vino aquí el número que yo necesito.Entonces, este valor aquí nada más es qué? n. Y aquí, vamos a depositar en nuestra variable,n .Shift-Enter, tenemos N.Ahora, ¿qué está faltando también?Está faltando, por ejemplo, calcular K. Y como yo calculo K,k = 1 + (10/3) * np.log10(n), que es lo que nosotros tenemos aquí, por...Traemos nuestra librería o biblioteca en un pie, aquí en np, y le estamos diciendo, por favor, pase el logaritmo 10 aquí,y coloca nuestro tamaño de nuestra muestra, de nuestra base de datos.Entonces, ¿cuántas clases nos va a dar?Y aquí, O.Tenemos Shift-Enter de nuevo. Hemos depositado en una variable. Ahora queremos imprimir. K va a ser qué? Él nos está mandando a hacer 17 clases.

Entonces, para hacer estas 17 clases, como esto de aquí, que hemos hecho aquí de manera personalizada,¿dónde está?Aquí.Esta parte de aquí, nosotros hemos personalizado así.Pero él nos está sugiriendo que hagamos 17 de estos.Entonces, si él nos mandó a hacer 17, ¿cómo creamos esa tabla de frecuencia?Lo único que hacemos es, ya hemos hecho esto anteriormente, es traer este código de aquí.Vamos a copiar este código.Vamos a crear aquí.Y en vez de ser la variable frecuencia, vamos a llamar frecuencia_sturges.Y aquí tenemos pd.value_counts(pd.cut( x = datos.Ingreso, bins = clases, labels = label, include_lowest = True)), solo que en bins va a ser 17. Y en la parte que es Labels, no vamos a tener label porque básicamente no es personalizado.Apenas tenemos los bins, que serían las clases. Y, de nuevo, nosotros estamos incluyendo el valor menor, que va a ser cero. Entonces, que vaya de cero al valor de 200.000, que era el salario máximo.Shift, Enter.Y, voilà.Tenemos aquí Ingreso.¿Cuánta es la frecuencia que nosotros tenemos?Y, básicamente, nos está mostrando, no se preocupen que aquí está en negativo,porque es de 200.000 a 11.764, 11.764 y por delante.Entonces, él está haciendo de manera mucho más detallada lo que nosotros hemos hecho de manera personalizada.Esto es una manera también que se puede usar.Sin embargo, si ustedes ya conocen su base de datos,ustedes pueden hacer algo más personalizado para que no quede en este entendimiento.Entonces, lo que yo sugiero también es personalizar.Ustedes van a tener que renombrar esto de aquí, como ya hemos visto en las clases pasadas también.Y es esto.

Ahora, ¿cuál es el ejercicio que yo les propongo que ustedes hagan aquí?¿Está bien?El ejercicio que yo les propongo es que ya hemos creado aquí nuestra tabla de frecuencia.¿Correcto? De valores absolutos. De nuevo, ¿qué es esto lo que significa?Y o quiero ver la proporción por cada una de las clases. Entonces, el ejercicio que ustedes van a tener que hacer es crear este mismo ejercicio,solo que ahora trayendo los porcentajes.¿Está bien? Entonces, hemos visto cómo crear las tablas de frecuencia para nuestras variables cuantitativas. Y en las próximas clases vamos a ver ahora cómo representar esto de manera un poco más visual, un poco más gráfica. ¿Está bien? Entonces, nos vemos en la próxima clase.Chao, chao.

In [74]:
porcentaje_sturges = pd.value_counts(
pd.cut(x = datos.Ingreso,
                    bins = 17,
                    include_lowest = True),
                    sort=False,normalize=True).round(4)*100

porcentaje_sturges

/tmp/ipython-input-3816370228.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  porcentaje_sturges = pd.value_counts(


Ingreso
(-200.001, 11764.706]       98.38
(11764.706, 23529.412]       1.33
(23529.412, 35294.118]       0.22
(35294.118, 47058.824]       0.02
(47058.824, 58823.529]       0.02
(58823.529, 70588.235]       0.01
(70588.235, 82352.941]       0.01
(82352.941, 94117.647]       0.00
(94117.647, 105882.353]      0.01
(105882.353, 117647.059]     0.00
(117647.059, 129411.765]     0.00
(129411.765, 141176.471]     0.00
(141176.471, 152941.176]     0.00
(152941.176, 164705.882]     0.00
(164705.882, 176470.588]     0.00
(176470.588, 188235.294]     0.00
(188235.294, 200000.0]       0.00
Name: proportion, dtype: float64

In [75]:
dist_frec_cuantitativa_amplitud_fija = pd.DataFrame({'Frecuencia':frecuencia_sturges,
  'Porcentaje':porcentaje_sturges})
dist_frec_cuantitativa_amplitud_fija

Frecuencia  Porcentaje
Ingreso                                         
(-200.001, 11764.706]          75594       98.38
(11764.706, 23529.412]          1022        1.33
(23529.412, 35294.118]           169        0.22
(35294.118, 47058.824]            19        0.02
(47058.824, 58823.529]            16        0.02
(58823.529, 70588.235]             5        0.01
(70588.235, 82352.941]             4        0.01
(82352.941, 94117.647]             1        0.00
(94117.647, 105882.353]            6        0.01
(105882.353, 117647.059]           0        0.00
(117647.059, 129411.765]           1        0.00
(129411.765, 141176.471]           0        0.00
(141176.471, 152941.176]           0        0.00
(152941.176, 164705.882]           0        0.00
(164705.882, 176470.588]           0        0.00
(176470.588, 188235.294]           0        0.00
(188235.294, 200000.0]             3        0.00

## <font color=#56B4E9>2.4 Histograma</font>
***

El <b>HISTOGRAMA</b> es la representación gráfica de una distribución de frecuencia. Es un gráfico formado por un conjunto de rectángulos colocados uno al lado del otro, donde el área de cada rectángulo es proporcional a la frecuencia de la clase que representa.

### Importando la biblioteca

https://seaborn.pydata.org/

# <font color=#56B4E9>3 MEDIDAS DE TENDENCIA CENTRAL</font>
***

## DataFrame de ejemplo

In [ ]:
df = pd.DataFrame(data = {'María': [8, 10, 4, 8, 6, 10, 8],
                          'Pedro': [10, 2, 0.5, 1, 3, 9.5, 10],
                          'Pablo': [7.5, 8, 7, 8, 8, 8.5, 7]},
                  index = ['Matemática',
                           'Portugués',
                           'Inglés',
                           'Geografía',
                           'Historia',
                           'Física',
                           'Química'])
df.rename_axis('Asignaturas', axis = 'columns', inplace = True)
df

Asignaturas  María  Pedro  Pablo
Matemática       8   10.0    7.5
Portugués       10    2.0    8.0
Inglés           4    0.5    7.0
Geografía        8    1.0    8.0
Historia         6    3.0    8.0
Física          10    9.5    8.5
Química          8   10.0    7.0

## <font color=#56B4E9>3.1 Media aritmética</font>
***

Es representada por $\mu$ cuando se refiere a la población y por $\bar{X}$ cuando se refiere a la muestra

# $$\mu = \frac 1n\sum_{i=1}^{n}X_i$$

donde

$n$ = número de observaciones (registros)

$X_i$ = valor de la i-ésima observación (registro)

## <font color=#56B4E9>3.2 Mediana</font>
***

Para obtener la mediana de un conjunto de datos, debemos proceder de la siguiente manera:
1. Ordenar el conjunto de datos;
2. Identificar el número de observaciones (registros) del conjunto de datos ($n$);
3. Identicar el elemento del medio:

> Cuando $n$ sea impar, la posición del elemento del medio se obtendrá de la siguiente manera:


# $$Elemento_{Md} = \frac{n+1}{2}$$

> Cuando $n$ sea par, la posición del elemento mediano se obtendrá de la siguiente manera:


# $$Elemento_{Md} = \frac{n}2$$

4. Obtener la mediana:

> Cuando $n$ sea impar:


# $$Md = X_{Elemento_{Md}}$$

> Cuando $n$ sea par:


# $$Md = \frac{X_{Elemento_{Md}} + X_{Elemento_{Md}+1}}2$$
***

### Ejemplo 1 - n impar

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img002.png' width='40%' style="float:left">

### Ejemplo 2 - n par

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img003.png' width='50%' style="float:left">

### Obtener la mediana en nuestro conjunto de datos

## <font color=#56B4E9>3.3 Moda</font>
***

La moda se puede definir como el valor más frecuente de un conjunto de datos. La moda es ampliamente utilizada para datos cualitativos.

### Obteniendo la moda de nuestro dataset

## <font color=#56B4E9>3.4 Relación entre media, mediana e moda</font>
***

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img004.png' width='80%'>

### Evaluando la variable INGRESO

***

### Evaluando la variable ALTURA

***

### Evaluando la variable AÑOS DE ESTUDIO

# <font color=#56B4E9>4 MEDIDAS DE LOCALIZACIÓN</font>
***

## <font color=#56B4E9>4.1 Cuartiles, deciles y percentiles</font>
***

Hay una serie de medidas de posición similares en su diseño a la mediana, aunque no son medidas de tendencia central. Como se sabe, la mediana divide la distribución en dos partes iguales en términos del número de elementos en cada parte. Los cuartiles, por otro lado, permiten que la distribución se divida en cuatro partes iguales en términos del número de elementos en cada uno; deciles en diez partes y centiles en cien partes iguales.

## <font color=#56B4E9>4.2 Box-plot</font>
***

El *box plot* da una idea de la posición, dispersión, asimetría, colas y valores extremos. La posición central está dada por la mediana y la dispersión por $IIQ$ (distancia inter cuartil). Las posiciones relativas de $Q1$, $Mediana$ y $Q3$ dan una idea de la simetría de la distribución. Las longitudes de las colas están dadas por las líneas que van desde el rectángulo a los valores remotos y por los valores extremos.

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img005.png' width='65%'>

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img006.png' width='80%'>

# <font color=#56B4E9>5 MEDIDAS DE DISPERSIÓN</font>
***

Aunque las medidas de tendencia central y de localización proporcionan un resumen muy importante de los datos, pueden no ser suficientes para caracterizar diferentes conjuntos, especialmente cuando las observaciones de una distribución dada presentan datos muy dispersos.

## <font color=#56B4E9>5.1 Desviación media absoluta</font>
***


# $$DM = \frac 1n\sum_{i=1}^{n}|X_i-\bar{X}|$$


## <font color=#56B4E9>5.2 Varianza</font>
***

### Varianza

La varianza se construye a partir de las diferencias entre cada observación y la media de los datos, es decir, la desviación alrededor de la media. Al calcular la varianza, las desviaciones alrededor de la media son elevadas al cuadrado.

### Varianza de la población

# $$\sigma^2 = \frac 1n\sum_{i=1}^{n}(X_i-\mu)^2$$

### Varianza de la muestra

# $$S^2 = \frac 1{n-1}\sum_{i=1}^{n}(X_i-\bar{X})^2$$

## <font color=#56B4E9>5.3 Desviación estándar</font>
***

Una de las restricciones de la varianza es el hecho de que proporciona medidas cuadráticas de las unidades originales; la varianza de las medidas de longitud, por ejemplo, está en unidades de área. Por lo tanto, el hecho de que las unidades sean diferentes dificulta la comparación de la dispersión con las variables que la definen. Una forma de eliminar esta dificultad es considerar su raíz cuadrada.

### Desviación estándar de la población

# $$\sigma = \sqrt{\frac 1n\sum_{i=1}^{n}(X_i-\mu)^2} \Longrightarrow \sigma = \sqrt{\sigma^2}$$


### Desviación estándar de la muestra

# $$S = \sqrt{\frac 1{n-1}\sum_{i=1}^{n}(X_i-\bar{X})^2} \Longrightarrow S = \sqrt{S^2}$$